src/control/server/ctl_storage_rpc.go

```diff
@@ -184,19 +190,46 @@ func (c *ControlService) doFormat(i *IOServerInstance, resp *pb.StorageFormatRes
 		c.scm.formatted = true
 	}
 
+	var formatFailed bool
+
 	// scaffolding
+	bdevConfig, err := i.bdevConfig()
+	if err != nil {
+		return err
+	}
 	ctrlrResults := types.NvmeControllerResults{}
-	c.nvme.Format(i.runner.Config.Storage.Bdev, &ctrlrResults)
-	resp.Crets = ctrlrResults
+	// A config with SCM and no block devices is valid, apparently.
+	if len(bdevConfig.DeviceList) > 0 {
+		c.nvme.Format(bdevConfig, &ctrlrResults)
+		resp.Crets = ctrlrResults
+		formatFailed = ctrlrResults.HasErrors()
+	}
 
+	scmConfig, err := i.scmConfig()
+	if err != nil {
+		return err
+	}
 	mountResults := types.ScmMountResults{}
-	c.scm.Format(i.runner.Config.Storage.SCM, &mountResults)
+	c.scm.Format(scmConfig, &mountResults)
 	resp.Mrets = mountResults
+	formatFailed = formatFailed || mountResults.HasErrors()
+
+	c.log.Debugf("nvme formatted: %t, scm formatted: %t, has superblock: %t",
+		c.nvme.formatted, c.scm.formatted, hasSuperblock)
 
-	if !hasSuperblock && c.nvme.formatted && c.scm.formatted {
-		c.log.Debugf("storage format successful on server %d\n", i.runner.Config.Index)
+	if c.nvme.formatted && c.scm.formatted {
+		// Use this an indicator for whether storage format was requested and completed
+		// vs. storage was already formatted and skipped.
+		// TODO: Rework this logic to be less convoluted.
+		if !hasSuperblock {
+			c.log.Infof("storage format successful on server %d\n", i.runner.Config.Index)
+		}
+	}
+
+	// Only notify that storage is ready if there were no errors.
+	if !formatFailed {
```

Comment:
In the case where there are no bdev devices configured, c.nvme.formatted will be false.

Reply:
should we be setting that true if there are no devices? would that make the logic clearer?

Reply:
I think I'd prefer not to add more overloading to the meaning. It should only be set to true if something was formatted, IMO. In this case we already have a bit of overloading in that it will be true if the storage was either formatted now or was detected to have been formatted previously (as indicated by the presence of a valid superblock).